### ISSUE TITLE
PEP 695 syntax doesn't work for generic NamedTuples

### ISSUE THREAD

**Issue by AlexWaygood** [python/cpython#104616]
Since Python 3.11, defining generic `NamedTuple`s has been legal, like so:

```pycon
>>> from typing import *
>>> T = TypeVar("T")
>>> class Foo(NamedTuple, Generic[T]):
...     x: T
>>>
```

However, using the new PEP-695 syntax for generic `NamedTuple`s fails with a slightly inexplicable error message:

```pycon
>>> from typing import *
>>> class Foo[T](NamedTuple):
...     x: T
...
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<stdin>", line 1, in <generic parameters of Foo>
  File "C:\Users\alexw\coding\cpython\Lib\typing.py", line 2707, in __new__
    setattr(nm_tpl, key, ns[key])
AttributeError: attribute '__type_params__' of 'type' objects is not writable
```

Using PEP-695 syntax for generic `TypedDict`s seems to work fine, but isn't tested anywhere in `test_typing.py`, as far as I can tell:

```pycon
>>> class Bar[T](TypedDict):
...     x: T
...
>>> Bar.__type_params__
(T,)
>>> Bar.__orig_bases__
(<function TypedDict at 0x0000017E9A17D9D0>, typing.Generic[T])
>>> Bar.__mro__
(<class '__main__.Bar'>, <class 'typing.Generic'>, <class 'dict'>, <class 'object'>)
```

Cc. @JelleZijlstra

**Comment by AlexWaygood** [comment]
I guess we need to add `__type_params__` to this set here? https://github.com/python/cpython/blob/cfa517d5a68bae24cbe8d9fe6b8e0d4935e507d2/Lib/typing.py#L2673

**Comment by JelleZijlstra** [comment]
Thanks! I'll take a look tonight (though feel free to propose a solution yourself if you can get to it earlier). Adding to `_special` feels like it's not enough, because that will just skip the assignment, and we still want the `__type_params__` to end up on the generated class somehow.

**Comment by AlexWaygood** [comment]
> Thanks! I'll take a look tonight (though feel free to propose a solution yourself if you can get to it earlier). Adding to `_special` feels like it's not enough, because that will just skip the assignment, and we still want the `__type_params__` to end up on the generated class somehow.

Ah, I didn't realise that it wasn't possible to set the `__type_params__` field on class objects. Maybe we should allow setting it on class objects? I don't see a strong reason for it to be read-only; this case would benefit from it; and in general I feel like you should be able to mutate most fields on class objects

**Comment by JelleZijlstra** [comment]
Yes, it probably should be writable; I thought it already was. I can make that change.

**Comment by JelleZijlstra** [comment]
Fixed in #104634
